### ISSUE TITLE
RS^3Mamba：训练和测试过程中遇见的fixture 'net' not found错误

### ISSUE THREAD

**Issue by WangTianYi0102** [sstary/SSRS#19]
你好，我在训练时遇见了下面这个报错导致无法完成
def test(net, test_ids, all=False, stride=WINDOW_SIZE[0], batch_size=BATCH_SIZE, window_size=WINDOW_SIZE):
E       fixture 'net' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
它提示我说测试用的net无法被找到，但是我已经确认修改了网络的调用，下面是报错信息和报错的代码位置，如果方便的话，麻烦您提供帮助
![1713601256164](https://github.com/sstary/SSRS/assets/163794144/bf247e56-a145-4c9c-a0d2-1b21efed8537)
![1713601429053](https://github.com/sstary/SSRS/assets/163794144/3a71e332-a5ce-4a39-bac6-f7272fddc2e0)

**Comment by WangTianYi0102** [comment]
import numpy as np
from glob import glob
from tqdm import tqdm_notebook as tqdm
from sklearn.metrics import confusion_matrix
import random, time
import itertools
import matplotlib.pyplot as plt
import torch
import torch.nn.functional as F
import torch.utils.data as data
import torch.optim as optim
import torch.optim.lr_scheduler
import torch.nn.init
from utils_Mamba import *
from torch.autograd import Variable
from IPython.display import clear_output
from model.UNetFormer import UNetFormer
from model.RS3Mamba import RS3Mamba, load_pretrained_ckpt

try:
    from urllib.request import URLopener
except ImportError:
    from urllib import URLopener

if MODEL == 'UNetformer':
    net = UNetFormer(num_classes=N_CLASSES).cuda()
elif MODEL == 'RS3Mamba':
    net = RS3Mamba(num_classes=N_CLASSES).cuda()
    net = load_pretrained_ckpt(net)

params = 0
for name, param in net.named_parameters():
    params += param.nelement()
print(params)

# for name, parms in net.named_parameters():
#     print('%-50s' % name, '%-30s' % str(parms.shape), '%-10s' % str(parms.nelement()))

# Load the datasets
print("training : ", str(len(train_ids)) + ", testing : ", str(len(test_ids)) + ", Stride_Size : ", str(Stride_Size), ", BATCH_SIZE : ", str(BATCH_SIZE))
train_set = ISPRS_dataset(train_ids, cache=CACHE)
train_loader = torch.utils.data.DataLoader(train_set,batch_size=BATCH_SIZE)

base_lr = 0.01
params_dict = dict(net.named_parameters())
params = []
for key, value in params_dict.items():
    if '_D' in key:
        # Decoder weights are trained at the nominal learning rate
        params += [{'params':[value],'lr': base_lr}]
    else:
        # Encoder weights are trained at lr / 2 (we have VGG-16 weights as initialization)
        params += [{'params':[value],'lr': base_lr / 2}]

optimizer = optim.SGD(net.parameters(), lr=base_lr, momentum=0.9, weight_decay=0.0005)
# We define the scheduler
scheduler = optim.lr_scheduler.MultiStepLR(optimizer, [25, 35, 45], gamma=0.1)
# scheduler = optim.lr_scheduler.MultiStepLR(optimizer, [15, 25, 35], gamma=0.1)

def test(net, test_ids, all=False, stride=WINDOW_SIZE[0], batch_size=BATCH_SIZE, window_size=WINDOW_SIZE):
    # Use the network on the test set
    test_images = (1 / 255 * np.asarray(io.imread(DATA_FOLDER.format(id)), dtype='float32') for id in test_ids)
    test_labels = (np.asarray(io.imread(LABEL_FOLDER.format(id)), dtype='uint8') for id in test_ids)
    if DATASET == 'Urban':
        eroded_labels = ((np.asarray(io.imread(ERODED_FOLDER.format(id)), dtype='int64') - 1) for id in test_ids)
    else:
        eroded_labels = (convert_from_color(io.imread(ERODED_FOLDER.format(id))) for id in test_ids)
    all_preds = []
    all_gts = []
    # Switch the network to inference mode
    with torch.no_grad():
        for img, gt, gt_e in tqdm(zip(test_images, test_labels, eroded_labels), total=len(test_ids), leave=False):
            pred = np.zeros(img.shape[:2] + (N_CLASSES,))

            total = count_sliding_window(img, step=stride, window_size=window_size) // batch_size
            for i, coords in enumerate(
                    tqdm(grouper(batch_size, sliding_window(img, step=stride, window_size=window_size)), total=total,
                        leave=False)):
                # Build the tensor
                image_patches = [np.copy(img[x:x + w, y:y + h]).transpose((2, 0, 1)) for x, y, w, h in coords]
                image_patches = np.asarray(image_patches)
                image_patches = Variable(torch.from_numpy(image_patches).cuda(), volatile=True)

                # Do the inference
                outs = net(image_patches)
                outs = outs.data.cpu().numpy()

                # Fill in the results array
                for out, (x, y, w, h) in zip(outs, coords):
                    out = out.transpose((1, 2, 0))
                    pred[x:x + w, y:y + h] += out
                del (outs)

            pred = np.argmax(pred, axis=-1)
            all_preds.append(pred)
            all_gts.append(gt_e)
            clear_output()
            
    accuracy = metrics(np.concatenate([p.ravel() for p in all_preds]),
                       np.concatenate([p.ravel() for p in all_gts]).ravel())
    if all:
        return accuracy, all_preds, all_gts
    else:
        return accuracy


def train(net, optimizer, epochs, scheduler=None, weights=WEIGHTS, save_epoch=1):
    weights = weights.cuda()

    iter_ = 0
    MIoU_best = 0.80
    for e in range(1, epochs + 1):
        if scheduler is not None:
            scheduler.step()
        net.train()
        for batch_idx, (data, target) in enumerate(train_loader):
            data, target = Variable(data.cuda()), Variable(target.cuda())
            optimizer.zero_grad()
            output = net(data)
            loss = loss_calc(output, target, weights)

            loss.backward()
            optimizer.step()

            # losses[iter_] = loss.data
            # mean_losses[iter_] = np.mean(losses[max(0, iter_ - 100):iter_])

            if iter_ % 100 == 0:
                clear_output()
                pred = np.argmax(output.data.cpu().numpy()[0], axis=0)
                gt = target.data.cpu().numpy()[0]
                print('Train (epoch {}/{}) [{}/{} ({:.0f}%)]\tLr: {:.6f}\tLoss: {:.6f}\tAccuracy: {}'.format(
                    e, epochs, batch_idx, len(train_loader),
                    100. * batch_idx / len(train_loader), optimizer.state_dict()['param_groups'][0]['lr'], loss.data, accuracy(pred, gt)))
            iter_ += 1

            del (data, target, loss)

            ## !!! You can increase the frequency of testing to find better models.
            if iter_ % 500 == 0:
                # We validate with the largest possible stride for faster computing
                net.eval()
                MIoU = test(net, test_ids, all=False, stride=Stride_Size)
                net.train()
                if MIoU > MIoU_best:
                    if DATASET == 'Vaihingen':
                        torch.save(net.state_dict(), '/data/AIfusion/liuzhongyun/Datasets/Vaihingen/resultsv/{}_epoch{}_{}'.format(MODEL, e, MIoU))
                    elif DATASET == 'Urban':
                        torch.save(net.state_dict(), './resultsu/{}_epoch{}_{}'.format(MODEL, e, MIoU))
                    MIoU_best = MIoU

if MODE == 'Train':
    train(net, optimizer, 100, scheduler)
elif MODE == 'Test':
    if DATASET == 'Vaihingen':
        net.load_state_dict(torch.load('/data/AIfusion/liuzhongyun/Datasets/Vaihingen/resultsv/RS3Mamba_epoch37_0.8200420260316154'), strict=False) # seg
        net.eval()
        MIoU, all_preds, all_gts = test(net, test_ids, all=True, stride=32)
        print("MIoU: ", MIoU)
        for p, id_ in zip(all_preds, test_ids):
            img = convert_to_color(p)
            # plt.imshow(img) and plt.show()
            io.imsave('/data/AIfusion/liuzhongyun/Datasets/Vaihingen/resultsv/Test_Result/inference8278_'+MODEL+'_tile_{}.png'.format(id_), img)

    elif DATASET == 'Urban':
        net.load_state_dict(torch.load('./resultsu/UNetformer_Embed_epoch23_0.5055703729081706'), strict=False)  # seg
        net.eval()
        MIoU, all_preds, all_gts = test(net, test_ids, all=True, stride=32)
        print("MIoU: ", MIoU)
        for p, id_ in zip(all_preds, test_ids):
            img = convert_to_color(p)
            # plt.imshow(img) and plt.show()
            io.imsave('./resultsu/inference5058_'+MODEL+'_tile_{}.png'.format(id_), img)

**Comment by sstary** [comment]
你好，你的问题可能和函数名有关，请参考：https://blog.csdn.net/weixin_39504171/article/details/106276568。如果该方法无效，请将你的项目文件打包发我，你应该对项目结构做了调整，我需要自己跑一下才能找到原因。

**Comment by WangTianYi0102** [comment]
您好，已经解决该问题，感谢您百忙之中的提点！

**Comment by sstary** [comment]
如果项目对你有帮助，请给我们一个star
